### DISCUSSION TITLE
Two instances same database record

### DISCUSSION THREAD

**Discussion by nicolaszein** [sqlalchemy/sqlalchemy#5986] [Usage Questions] [answered]
Is there a way of having two instances of the same database record?

I'm asking because I'm trying to follow the hexagonal architecture by decoupling the domain from external dependencies such as the database. 

So, imagine that we have a users table represented by the model:
```py
class UserModel(Base):
    __tablename__ = 'user'

    id = Column(Integer, primary_key=True)
    name = Column(String)
    fullname = Column(String)
```
And we have our pure domain:
```py
@dataclass
class User:
    name: str
    full_name: str
```

That way I can do all my domain's operations and then convert to the model to be able to persist.
So in a scenario where I need to get a user to update the name, it would look something like this:
```
user_model = User.get(id)
user_domain = user_model.to_domain()
user_domain.name = 'new_name'
new_user_model = user_domain.to_model()
```
It's a silly example I know 😅 , but it is just to give an idea.

When I tried to persist it raises:
`FlushError: New instance with identity key conflicts with persistent instance...`
 
And that happens because when I got the user for the first time the pk was persisted on the session's identity map.
And because I built a new instance with the same pk it returns this Exception.

I would like to know if there is some workaround or some way to accomplish that.

Thank you very much in advance! 😄 

**Have a nice day!**

**Comment by zzzeek** [comment, marked as answer]
yeah so the method you want to use when you have a new object and you want to merge your changes back into a session where that object is already present is the merge() method: https://docs.sqlalchemy.org/en/13/orm/session_state_management.html#merging

when you call merge(), get the return value from it, that's the object that's actually in the session.

another way you could do this, would be to expunge() the old object first before putting your new one in the session.   the problem with that is if you havea whole graph of objects loaded for that session, your target object might be referenced by other objects, and those references would stay around unless you expired everything.

I would think, if I were trying to do this design, that I'd use a coarse grained approach, which is to load the full graph within the scope of a Session, then to_domain() the whole thing in one go, then do all your business logic with your dataclass domain model, then go back to to_model() and populate the Session completely clean, that is, do an expunge_all() and re-add() everything from scratch.  that way there's more of a straight boundary between the two sections.

**Reply by nicolaszein** [reply]
Hi @zzzeek, how are you?

First of all, thank you so much for your help and time... I really appreciated 😃 

So essentially, what I should do is:
```py
user_model = session.query(UserModel).filter_by(id=id)
user_domain = user_model.to_domain()
user_domain.name = 'new_name'
session.expunge_all()
new_user_model = user_domain.to_model()
session.add(new_user_model)
session.commit()
```

Am I right?

**Reply by nicolaszein** [reply]
I tried this way and it stopped the FlushError exception from being raised. However, now it is raising IntegrityError because of the same id, sqlalchemy thinks that I'm adding a new user, not updating. 😞 

I guess there is no way for doing this right?

**Reply by zzzeek** [reply]
so your to_model() is making a new instance, which by default looks like it's [transient](https://docs.sqlalchemy.org/en/13/glossary.html#term-transient), meaning, it's never been INSERTed.  there's no problem here, you can make it behave as [detached](https://docs.sqlalchemy.org/en/13/glossary.html#term-detached) instead using [make_transient_to_detached](https://docs.sqlalchemy.org/en/13/orm/session_api.html?highlight=transient_to_detached#sqlalchemy.orm.session.make_transient_to_detached).   then add() it, then it will look like it's an UPDATE.

**Reply by nicolaszein** [reply]
Nice!!!

That really worked...
Thank you @zzzeek so much... 😃
